ddtrace/internal/writer/writer.py

```diff
@@ -77,6 +80,29 @@ class NoEncodableSpansError(Exception):
 DEFAULT_SMA_WINDOW = 10
 
 
+def make_weak_method_hook(bound_method):
+    """
+    Wrap a bound method so that it is called via a weakref to its instance.
+    If the instance has been garbage-collected, the hook is a no-op.
+    """
+    if not hasattr(bound_method, "__self__") or bound_method.__self__ is None:
+        raise TypeError("make_weak_method_hook expects a bound method")
```

Comment:
In general we avoid raising exceptions in customer apps. Ideally when the ddtrace library fails it should should just log a warning an exit gracefully.